src/algorithms/local_search/local_search.cpp

```diff
@@ -113,6 +113,173 @@ LocalSearch<Route,
+          continue;
+      }
+
+      // Build replacement sequence for current insertion.
+      std::vector<Index> modified_with_pd({j});
+      Amount modified_delivery = input.zero_amount();
+
+      for (Index delivery_r = pickup_r; delivery_r <= route.size();
+              ++delivery_r) {
+          // Update state variables along the way before potential
+          // early abort.
+          if (pickup_r < delivery_r) {
+              modified_with_pd.push_back(route.route[delivery_r - 1]);
+              const auto& new_modified_job =
+                  input.jobs[route.route[delivery_r - 1]];
+              if (new_modified_job.type == JOB_TYPE::SINGLE) {
+                  modified_delivery += new_modified_job.delivery;
+              }
+          }
+
+          if (!(bool)valid_delivery_insertions[delivery_r]) {
+              continue;
+          }
+
+          Gain pd_cost;
+          if (pickup_r == delivery_r) {
+              pd_cost = utils::addition_cost(input,
+                      matrix,
+                      j,
+                      v_target,
+                      route.route,
+                      pickup_r,
+                      pickup_r + 1);
+          } else {
+              pd_cost = p_add + d_adds[delivery_r];
+          }
+
+          // Normalize cost per job for consistency with single jobs.
+          Gain current_cost =
+              static_cast<Gain>(static_cast<double>(pd_cost) / 2);
+
+          if (current_cost < result.cost) {
+              modified_with_pd.push_back(j + 1);
+
+              // Update best cost depending on validity.
+              bool is_valid =
+                  route
+                  .is_valid_addition_for_capacity_inclusion(input,
+                          modified_delivery,
+                          modified_with_pd
+                          .begin(),
+                          modified_with_pd
+                          .end(),
+                          pickup_r,
+                          delivery_r);
+
+              is_valid =
+                  is_valid &&
+                  route.is_valid_addition_for_tw(input,
+                          modified_with_pd.begin(),
+                          modified_with_pd.end(),
+                          pickup_r,
+                          delivery_r);
+
+              modified_with_pd.pop_back();
+
+              if (is_valid) {
+                  result = {current_cost, 0, pickup_r, delivery_r};
+              }
+          }
+      }
+  }
+  return result;
+}
+
+template <class Route>
+RouteInsertion compute_best_insertion(const Input& input, const Index j, Index v, const Route& route) {
+
+  const auto& current_job = input.jobs[j];
+  if (current_job.type == JOB_TYPE::SINGLE) {
+      return compute_best_insertion_single(input, j, v, route);
+  } else if (current_job.type == JOB_TYPE::PICKUP) {
+      return compute_best_insertion_pd(input, j, v, route);
+  } else {
+    return empty_insert;
```

Comment:
If I'm correct, this is never used, except if the caller incorrectly filters out `JOB_TYPE::DELIVERY`? If so it could make sense here to assert that the type is either `JOB_TYPE::SINGLE` or `JOB_TYPE::PICKUP`, then the `if`s could boil down to a simple ternary operator.